### PR TITLE
added scrolling handlers of ScrollArrangements & Chinese Translation of it [to branch ucr]

### PR DESCRIPTION
**Pull request by ColinTree**
- Add events and methods that listens to & control the ScrollView (for both VerticalScrollArrangement and HorizontalScrollArrangement)
- Component version upgraded (1->2 no block need to upgrade)
- Chinese Simplifed Translation for these is ready

### CODE REVIEW

**Comment by ColinTree** [comment]
feel like every thing is ready now :P

**Comment by ColinTree** [comment]
hello @ewpatton 
i wanna ask something about github PR system: 
as a example, if i add some more commit here, will them be merged into one commit (this PR) when it is approved to be merged into the main repo?
cause i got something new and wanna add them.

**Comment by ewpatton** [comment]
@weihuali0509 Would you mind checking the Chinese translations?

File: appinventor/components/src/com/google/appinventor/components/runtime/HorizontalScrollArrangement.java [hunk 1]
```diff
@@ -7,28 +7,162 @@
...
+import com.google.appinventor.components.annotations.SimpleEvent;
+import com.google.appinventor.components.annotations.SimpleProperty;
+import com.google.appinventor.components.annotations.PropertyCategory;
 import com.google.appinventor.components.common.ComponentCategory;
 import com.google.appinventor.components.common.ComponentConstants;
 import com.google.appinventor.components.common.YaVersion;
 
+import android.view.View;
+import android.widget.HorizontalScrollView;
+import android.view.ViewTreeObserver.*;
```
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/runtime/HorizontalScrollArrangement.java]: We prefer specific imports over wildcards.

File: appinventor/components/src/com/google/appinventor/components/runtime/HorizontalScrollArrangement.java [hunk 2]
```diff
@@ -7,28 +7,162 @@
...
 /**
  * A horizontal arrangement of components
  * @author sharon@google.com (Sharon Perl)
  * @author jis@mit.edu (Jeffrey I. Schiller)
  *
+ * events and methods about scrolling
+ * @author 502470184@qq.com (ColinTree, YANG)
  */
-@DesignerComponent(version = YaVersion.HORIZONTALSCROLLARRANGEMENT_COMPONENT_VERSION,
+@DesignerComponent(version = YaVersion.HORIZONTALSCROLLARRANGEMENT_COMPONENT_VERSION,
```
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/runtime/HorizontalScrollArrangement.java]: Remove end-of-line whitespace.

File: appinventor/components/src/com/google/appinventor/components/runtime/HorizontalScrollArrangement.java [hunk 3]
```diff
@@ -7,28 +7,162 @@
...
     "scrollable.",
     category = ComponentCategory.LAYOUT)
 @SimpleObject
-public class HorizontalScrollArrangement extends HVArrangement {
-  public HorizontalScrollArrangement(ComponentContainer container) {
-    super(container, ComponentConstants.LAYOUT_ORIENTATION_HORIZONTAL,
-      ComponentConstants.SCROLLABLE_ARRANGEMENT);
-  }
+public class HorizontalScrollArrangement extends HVArrangement implements OnScrollChangedListener {
+    public HorizontalScrollArrangement(ComponentContainer container) {
```
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/runtime/HorizontalScrollArrangement.java]: Indentation should be 2 spaces, not 4.

File: appinventor/components/src/com/google/appinventor/components/runtime/VerticalScrollArrangement.java [hunk 4]
```diff
@@ -7,15 +7,25 @@
...
+import com.google.appinventor.components.annotations.SimpleEvent;
+import com.google.appinventor.components.annotations.SimpleProperty;
+import com.google.appinventor.components.annotations.PropertyCategory;
 import com.google.appinventor.components.common.ComponentCategory;
 import com.google.appinventor.components.common.ComponentConstants;
 import com.google.appinventor.components.common.YaVersion;
 
+import android.view.View;
+import android.widget.ScrollView;
+import android.view.ViewTreeObserver.*;
```
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/runtime/VerticalScrollArrangement.java]: See my comments on HorizontalScrollArrangement.java

**Comment by ColinTree** [comment]
Done @ewpatton 
I m thinking of what if i sepreate the translation and the others (may be all my PRs). Should I?

**Comment by ewpatton** [comment]
@ColinTree I think it is fine to keep the translations with the components they translate. Otherwise, we would have two PRs where one is dependent on another.

**Comment by cavedunissin** [comment]
Same idea with @ewpatton

File: appinventor/appengine/src/com/google/appinventor/client/OdeMessages.java [hunk 5]
```diff
@@ -3616,6 +3626,19 @@ String newerVersionComponentException(String componentType, int srcCompVersion,
...
   //Params
+  @DefaultMessage("scrollPosition")
+  @Description("")
+  String scrollPositionParams();
+
+  @DefaultMessage("position")
+  @Description("")
+  String positionParams();
+
+  @DefaultMessage("displacement")
```
Comment (ewpatton) [inline appinventor/appengine/src/com/google/appinventor/client/OdeMessages.java]: I think that `displacement` might be a term unknown to some of our younger users (~10 years old). I think it might be better for the related methods to use the term `amount` or `amountInPixels` (to indicate units), e.g. `ScrollBy(amount)`.

File: appinventor/appengine/src/com/google/appinventor/client/OdeMessages.java [hunk 6]
```diff
@@ -4345,6 +4368,28 @@ String newerVersionComponentException(String componentType, int srcCompVersion,
   String widthParams();
 
   //Events
+
+  @DefaultMessage("ReachLeftEnd")
```
Comment (ewpatton) [inline appinventor/appengine/src/com/google/appinventor/client/OdeMessages.java]: I'd like to think about this w.r.t. RTL support. In iOS, the terms used are leading and trailing edges where leading is left in LTR languages and right in RTL languages, and vice versa for trailing. We should look at how these events behave on devices configured with RTL languages. Do the semantics end up backward?

File: appinventor/appengine/src/com/google/appinventor/client/OdeMessages.java [hunk 7]
```diff
@@ -4746,6 +4791,71 @@ String newerVersionComponentException(String componentType, int srcCompVersion,
   String SensorValueChangedEvents();
 
   //Methods
+  @DefaultMessage("ArrowScrollLeftward")
```
Comment (ewpatton) [inline appinventor/appengine/src/com/google/appinventor/client/OdeMessages.java]: Ditto LTR vs RTL naming and behavior.

File: appinventor/components/src/com/google/appinventor/components/runtime/HorizontalScrollArrangement.java [hunk 8]
```diff
@@ -25,10 +35,129 @@
...
+
+  @Override
+  public void onScrollChanged() {
+    int scrollX = dx2px(getView().getScrollX());
+    if (scrollX < 0) {
+      scrollX = 0;
+    }
+    if (oldScrollX != scrollX) {
+      ScrollChanged(scrollX);
+      oldScrollX = scrollX;
```
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/runtime/HorizontalScrollArrangement.java]: I would update `oldScrollX` before dispatching the event. Otherwise, if there is a thrown exception in the handler the `oldScrollX` field might not be updated. Alternatively, wrap in a `try-finally` block.

File: appinventor/components/src/com/google/appinventor/components/runtime/HorizontalScrollArrangement.java [hunk 9]
```diff
@@ -25,10 +35,129 @@
...
+    }
+  }
+
+  @SimpleEvent(description = "Notice that if blocks like \"ScrollBy\" is excuted, this event would be called")
```
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/runtime/HorizontalScrollArrangement.java]: excuted => executed

File: appinventor/components/src/com/google/appinventor/components/runtime/HorizontalScrollArrangement.java [hunk 10]
```diff
@@ -25,10 +35,129 @@
...
+    if (scrollPosition == 0) {
+      ReachLeftEnd();
+    } else {
+      if (MaxScrollPosition() - ScrollPosition() <= 0) {
+        ReachRightEnd();
+      }
+    }
+  }
+
+  @SimpleEvent(description = "Notice that if blocks like \"ScrollLeftEnd\" is excuted, this event would be called")
```
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/runtime/HorizontalScrollArrangement.java]: Ditto

File: appinventor/components/src/com/google/appinventor/components/runtime/HorizontalScrollArrangement.java (hunk 9 above)
Comment (ewpatton) [reply appinventor/components/src/com/google/appinventor/components/runtime/HorizontalScrollArrangement.java]: I would rephrase this as "Runs when the scroll position of the arrangement changes." In general the event descriptions should complete the sentence "The <name> event block ..."

File: appinventor/components/src/com/google/appinventor/components/runtime/HorizontalScrollArrangement.java [hunk 11]
```diff
@@ -25,10 +35,129 @@
...
+  public void ReachLeftEnd() {
+    EventDispatcher.dispatchEvent(this, "ReachLeftEnd");
+  }
+
+  @SimpleEvent(description = "Notice that if blocks like \"ScrollRightEnd\" is excuted, this event would be called")
```
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/runtime/HorizontalScrollArrangement.java]: Ditto.

File: appinventor/components/src/com/google/appinventor/components/runtime/HorizontalScrollArrangement.java [hunk 12]
```diff
@@ -25,10 +35,129 @@
...
+    getScrollView().pageScroll(HorizontalScrollView.FOCUS_RIGHT);
+  }
+
+  @SimpleFunction(description = "Scroll to a specific position (without animation)")
+  public void ScrollTo(int position) {
+    getScrollView().scrollTo(px2dx(position), 0);
+  }
+
+  @SimpleFunction(description = "Scroll to a specific position (with animation)")
+  public void SmoothScrollTo(int position) {
```
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/runtime/HorizontalScrollArrangement.java]: Consider merging the `SmoothScroll...` and `Scroll...` methods by adding a new `animated` parameter that can be `true/false`.

File: appinventor/components/src/com/google/appinventor/components/runtime/VerticalScrollArrangement.java [hunk 13]
```diff
@@ -27,11 +37,129 @@
     "This version is scrollable",
     category = ComponentCategory.LAYOUT)
 @SimpleObject
-public class VerticalScrollArrangement extends HVArrangement {
+public class VerticalScrollArrangement extends HVArrangement implements OnScrollChangedListener {
```
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/runtime/VerticalScrollArrangement.java]: See my comments on HorizontalScrollArrangement and adapt accordingly.

File: appinventor/docs/reference/components/layout.html [hunk 14]
```diff
@@ -533,6 +533,10 @@ <h3>Properties</h3>
                       <dd></dd>
                       <dt><code>Image</code></dt>
                       <dd>Specifies the path of the component's image.  If there is both an Image and a BackgroundColor, only the Image will be visible.</dd>
+                      <dt><code>MaxScrollPosition</code></dt>
```
Comment (ewpatton) [inline appinventor/docs/reference/components/layout.html]: If you incorporate any of the suggestions above, update this file accordingly.

**Comment by ewpatton** [comment]
@cavedunissin @mintingle @weihuali0509 Would one of you mind reviewing the Chinese translations?

**Comment by ColinTree** [comment]
Plan to do:

- [ ] Find out the effect of RTL
- [ ] Add drop down selection to methods like ones in Clock: 
    image

**Comment by ColinTree** [comment]
@ewpatton I had tried in a emulator which the locale is set to Arabic. As I see in the System apps like settings, text does look like in RTL: 
image

But not thing changed in appinventor apps even the ListView, ListPicker and Spinner. I think it is because appinventor had set the text align.
image

And fortunately, events works as excepted: 
image
animation
